### PR TITLE
183 orchestrator tracks frontier cells

### LINKED ISSUE #158: orchestrator declares run as finished
- simulator switches to the next run when complete
- simulator exits when all run complete

### LINKED ISSUE #159: data collector puts completion time into logs

### LINKED ISSUE #182: add Frontier cells to UI

### LINKED ISSUE #183: orchestrator tracks frontier cells
frontier cells expand like elastic band around explored cells

### PR DESCRIPTION
**Pull request by Raa37**

### CODE REVIEW

File: Orchestrator.py [hunk 1]
```diff
@@ -143,87 +142,36 @@ def receive(self,frame):
...
         ]
 
+        # compute new frontiers
         for (cx, cy) in cellsExplored['cellsExplored']:
             for n in self._computeCellNeighbours(cx, cy):
                 if (
                     (n not in self.cellsExplored) and
                     (n not in self.cellsObstacle) and
-                    (n not in self.cellsUnreachable)
+                    self._checkForCornerFrontier(n) == False     # check that cell isn't a corner frontier
```
Comment (twatteyne) [inline Orchestrator.py]: rename method to `isCornerFrontier()`

File: Orchestrator.py [hunk 2]
```diff
@@ -438,4 +390,69 @@ def _computeCellNeighbours(self, cx, cy):
...
+                # at same connecting corner
+                for cell in connectedCells:
+                    if cell in self.cellsExplored:
+                        # remove cell from frontiers if it is already set as one
+                        if frontier in self.cellsFrontier:
+                            self.cellsFrontier.remove(frontier)
+                        returnVal = True
+                        break
+
         return returnVal
```
Comment (twatteyne) [inline Orchestrator.py]: add empty newline (please do from now on, it's the third time I have this comment)

File: Orchestrator.py [hunk 3]
```diff
@@ -438,4 +390,69 @@ def _computeCellNeighbours(self, cx, cy):
...
+        (cxmax, cymax)    = (cxmin + self.MINFEATURESIZE/2, cymin + self.MINFEATURESIZE/2)
+
+        topLeftCorner     = (cxmin, cymin)
+        topRightCorner    = (cxmax, cymin)
+        bottomLeftCorner  = (cxmin, cymax)
+        bottomRightCorner = (cxmax, cymax)
+
+        return [topLeftCorner, topRightCorner, bottomLeftCorner, bottomRightCorner]
+
+    def _checkForCornerFrontier(self, frontier):
```
Comment (twatteyne) [inline Orchestrator.py]: add a comment which specifies format of "frontier" (most probably not the right name)

**Review by twatteyne — CHANGES_REQUESTED** [review]
- add test cases for the function which computes whether it's a corner frontier
- what is the impact of this code on the speed of executation (give some idea)

**Comment by Raa37** [comment]
- This code increases execution time by x1.5

File: tests/test_isCornerFrontier.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    '''
+
+    orchestrator  = Orchestrator.Orchestrator(
+        numRobots = 1,
+        initX     = 1,
+        initY     = 1
+    )
+    orchestrator.MINFEATURESIZE = 1
+    orchestrator.cellsExplored = [(1,1), (1.5,1),(2,1),(5.5,1),(4,3),(1,4.5)]
+    orchestrator.cellsObstacle = [(1.0,0.5),(1.5,0.5),(5.5,0.5),(6,1),(4,2.5),(3.5,3),(2.5,3),(2,3.5),(0.5,4.5),(1,5)]
```
Comment (twatteyne) [inline tests/test_isCornerFrontier.py:93]: can you please draw up in ASCII art what this corresponds to?

**Review by twatteyne — CHANGES_REQUESTED** [review]
